src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.IncrementalMemberEditAnalyzer.cs

```diff
@@ -0,0 +1,367 @@
+                var memberSpans = members.SelectAsArray(member => member.FullSpan);
+                var changedMemberId = members.IndexOf(changedMember);
+                return (changedMember, changedMemberId, memberSpans, lastDocument);
+            }
+
+            private static async Task<ImmutableArray<DiagnosticData>> GetUpdatedDiagnosticsForMemberEditAsync(
+                ImmutableArray<DiagnosticData> diagnostics,
+                DocumentAnalysisData existingData,
+                DiagnosticAnalyzer analyzer,
+                DocumentAnalysisExecutor executor,
+                SyntaxNode changedMember,
+                int changedMemberId,
+                ImmutableArray<TextSpan> oldMemberSpans,
+                Func<DiagnosticAnalyzer, DocumentAnalysisExecutor, CancellationToken, Task<ImmutableArray<DiagnosticData>>> computeAnalyzerDiagnosticsAsync,
+                CancellationToken cancellationToken)
+            {
+                // We are performing semantic span-based analysis for member-only edit scenario.
+                // Instead of computing the analyzer diagnostics for the entire document,
+                // we have computed the new diagnostics just for the edited member span.
+                Debug.Assert(executor.AnalysisScope.Span.HasValue);
+                Debug.Assert(executor.AnalysisScope.Span.Value == changedMember.FullSpan);
+                Debug.Assert(diagnostics.All(d => !d.HasTextSpan || changedMember.FullSpan.IntersectsWith(d.GetTextSpan())));
+
+                // We now try to get the new document diagnostics by performing an incremental update:
+                //   1. Re-using all the old cached diagnostics outside the edited member node from a prior
+                //      document snapshot, but with updated diagnostic spans.
+                //      AND
+                //   2. Replacing old diagnostics for the edited member node in a prior document snapshot
+                //      with the new diagnostics for this member node in the latest document snaphot.
+                // If we are unable to perform this incremental diagnostics update,
+                // we fallback to computing the diagnostics for the entire document.
+                if (TryGetUpdatedDocumentDiagnostics(existingData, oldMemberSpans, diagnostics, changedMember.SyntaxTree, changedMember, changedMemberId, out var updatedDiagnostics))
+                {
+#if DEBUG_INCREMENTAL_ANALYSIS
+                    await ValidateMemberDiagnosticsAsync(executor, analyzer, updatedDiagnostics, cancellationToken).ConfigureAwait(false);
+#endif
+                    return updatedDiagnostics;
+                }
+                else
+                {
+                    // Incremental diagnostics update failed.
+                    // Fallback to computing the diagnostics for the entire document.
+                    var documentExecutor = executor.With(executor.AnalysisScope.WithSpan(null));
```

Comment:
is there anywhere we can log things like this in case we get reports of slow diagnostics?  Might help track things down later.

Reply:
Sure, but I want to avoid adding telemetry unless we actually see reports and know that someone is going to look at this data. We can certainly add such telemetry in future.